lib/routes/token.js

```diff
@@ -16,9 +19,146 @@ const P = require('../promise');
 const Scope = require('../scope');
 const validators = require('../validators');
 
+const HEX_STRING = validators.HEX_STRING;
+
 const MAX_TTL_S = config.get('expiration.accessToken') / 1000;
 const GRANT_AUTHORIZATION_CODE = 'authorization_code';
 const GRANT_REFRESH_TOKEN = 'refresh_token';
+const GRANT_JWT = 'urn:ietf:params:oauth:grant-type:jwt-bearer';
+
+const JWT_AUD = config.get('publicUrl') + '/v1/token';
+
+const SERVICE_CLIENTS = {};
+const JWTOOL = new JwTool(config.get('serviceClients').map(function(client) {
+  SERVICE_CLIENTS[client.jku] = client;
+  return client.jku;
+}));
+
+const PAYLOAD_SCHEMA = Joi.object({
+
+  client_id: validators.clientId
+    .when('grant_type', {
+      is: GRANT_JWT,
+      then: Joi.forbidden()
+    }),
+
+  client_secret: validators.clientSecret
+    .when('grant_type', {
+      is: GRANT_JWT,
+      then: Joi.forbidden()
+    }),
+
+  grant_type: Joi.string()
+    .valid(GRANT_AUTHORIZATION_CODE, GRANT_REFRESH_TOKEN, GRANT_JWT)
+    .default(GRANT_AUTHORIZATION_CODE)
+    .optional(),
+
+  ttl: Joi.number()
+    .max(MAX_TTL_S)
+    .default(MAX_TTL_S)
+    .optional(),
+
+  scope: Joi.alternatives().when('grant_type', {
+    is: GRANT_REFRESH_TOKEN,
+    then: Joi.string(),
+    otherwise: Joi.forbidden()
+  }),
+
+  code: Joi.string()
+    .length(config.get('unique.code') * 2)
+    .regex(validators.HEX_STRING)
+    .required()
+    .when('grant_type', {
+      is: GRANT_AUTHORIZATION_CODE,
+      otherwise: Joi.forbidden()
+    }),
+
+  refresh_token: validators.token
+    .required()
+    .when('grant_type', {
+      is: GRANT_REFRESH_TOKEN,
+      otherwise: Joi.forbidden()
+    }),
+
+  assertion: validators.assertion
+    .required()
+    .when('grant_type', {
+      is: GRANT_JWT,
+      otherwise: Joi.forbidden()
+    })
+
+});
+
+// No? Still want to press on? Well, OK. But you were warned.
+//
+// This route takes takes an authorization grant, and returns an
+// access_token if everything matches up.
+//
+// Steps from start to finish:
+//
+// 1. Confirm grant credentials.
+//    - If grant type is authorization code or refresh token, first
+//      confirm the client credentials in `confirmClient()`.
+//      - If grant type is authorization code, proceed to `confirmCode()`.
+//      - If grant type is refresh token, proceed to `confirmRefreshToken()`.
+//    - If grant type is a JWT, all information is in the JWT. So jump
+//      straight to `confirmJwt()`.
+// 2. Generate tokens.
+//   - An options object is passed to `generateTokens()`.
+//     - An access_token is generated.
+//     - If req.payload.access_type = 'offline', a refresh_token is also
+//       generated.
```

Comment:
It sounds like this will allow all grant types to generate refresh tokens.  I wonder if we should restrict it to only `authorization_code` clients.  Service clients have no legitimate use for a refresh token, since they can generate access tokens on demand, and having them able to generate refresh tokens could make it harder to e.g. thoroughly revoke service client access.